cmd/rehearse/main.go

```diff
@@ -226,25 +227,62 @@ func execute(jobs config.JobConfig, jobSpec *pjapi.ProwJobSpec, logger logrus.Fi
 		}
 	}
 
-	if len(rehearsals) > 0 {
-		if err := rehearsalConfigs.Create(); err != nil {
-			return fmt.Errorf("failed to prepare rehearsal ci-operator config ConfigMap: %v", err)
-		}
-		for _, job := range rehearsals {
-			created, err := submitRehearsal(job, jobSpec, logger, pjclient, dry)
-			if err != nil {
-				logger.WithError(err).Warn("Failed to execute a rehearsal presubmit presubmit")
-			} else {
-				logger.WithFields(pjutil.ProwJobFields(created)).Info("Submitted rehearsal prowjob")
-			}
-		}
-	} else {
+	if len(rehearsals) == 0 {
 		logger.Warn("No job rehearsals")
+		return nil
+	}
+	if err := rehearsalConfigs.Create(); err != nil {
+		return fmt.Errorf("failed to prepare rehearsal ci-operator config ConfigMap: %v", err)
+	}
+	watcher, err := pjclient.Watch(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("could not create watcher for ProwJobs: %v", err)
+	}
+	defer watcher.Stop()
+	pjs := make(map[string]struct{}, len(rehearsals))
+	for _, job := range rehearsals {
+		created, err := submitRehearsal(job, jobSpec, logger, pjclient, dry)
+		if err != nil {
+			logger.WithError(err).Warn("Failed to execute a rehearsal presubmit presubmit")
+		}
+		logger.WithFields(pjutil.ProwJobFields(created)).Info("Submitted rehearsal prowjob")
+		pjs[created.Name] = struct{}{}
+	}
+	success, err := waitForJobs(pjs, watcher.ResultChan(), logger)
+	if err != nil {
+		return fmt.Errorf("error while waiting for jobs: %s", err)
+	}
+	if !success {
+		return fmt.Errorf("not all ProwJobs were successful")
```

Comment:
I was briefly thinking about this, and I think it's worth the additional effort to think more about the failure modes here. We will probably need to easily distinguish between the "we submitted all rehearsals and some failed" (which is not actually a failure, but the rehearsal tool working as intended) and the cases where something actually failed within what the rehearsal attempts to do.

Reply:
Yes, that is merely a placeholder for now while I focus on the watch. Proper error handling will come later.